deps/wazuh_testing/wazuh_testing/process_resource_monitoring/README.md

```diff
@@ -0,0 +1,157 @@
+# Monitoring class for process data collection
+
+This package contains two public modules designed for monitoring and tracking purposes:
+
+## Monitor module
+
+The `monitor` module provides a class for monitoring processes and their child processes. Monitoring is conducted concurrently, with a separate thread for each process. The following data is collected for each scan:
+
+   - **Daemon**: daemon name.
+   - **Version**: version.
+   - **Timestamp**: timestamp of the scan.
+   - **PID**: PID of the process.
+   - **CPU**(%): CPU percentage of the process. It can exceed 100% if the process uses multiple threads.
+   - **VMS**: Virtual Memory Size.
+   - **RSS**: Resident Set Size.
+   - **USS**: Unique Set Size.
+   - **PSS**: Proportional Set Size.
+   - **SWAP**: Memory of the process in the swap space.
+   - **FD**: File descriptors opened by the process.
+   - **Read_Ops**: Read operations.
+   - **Write_Ops**: Write operations.
+   - **Disk_Read**: Bytes read by the process.
+   - **Disk_Written**: Bytes written by the process.
+
+## Disk Usage Tracker module
+
+The `disk_usage_tracker` module provides a class to monitor file and directory disk usage. The following data is collected for each file or directory:
+
+- **File:** Name of the file.
+- **Timestamp:** Timestamp of the scan.
+- **Path:** Full path of the file.
+- **Size:** Size in selected units of the file.
+- **Usage:** Percentage of the space the file takes relative to the partition's size.
+- **Mod_time:** Last time the file was modified.
+- **Acc_time:** Last time the file was accessed.
+- **Creat_time:** Creation time (Windows) or metadata change time (Unix).
+
+The package also provides the script `wazuh-metrics`, designed to interact with these modules. [More information](#script)
+ 
+### Process reference
+
+#### Wazuh manager
+
+| Process | Argument |
+| ------- | -------- |
+| wazuh-agentlessd | agentlessd |
+| wazuh-analysisd | analysisd |
+| wazuh_apid.py | apid |
+| wazuh-authd | authd |
+| wazuh_clusterd.py | clusterd |
+| wazuh-csyslogd | csyslogd |
+| wazuh-db | db |
+| wazuh-dbd | dbd |
+| wazuh-execd | execd |
+| wazuh-integratord | integratord |
+| wazuh-logcollector | logcollector |
+| wazuh-maild | maild |
+| wazuh-modulesd | modulesd |
+| wazuh-monitord | monitord |
+| wazuh-remoted | remoted |
+| wazuh-syscheckd | syscheckd |
```

Comment:
Some concerns arise from here:
- How would this be achieved? Since, as of right now and following the rationale of "making the tool Wazuh-independent" it should be able to monitor any "common" process, such as a python script (which is actually executed by the Python interpreter). In some cases like the aforementioned one the process can not be found in the name and has to be searched in the process arguments (in Unix-like systems you can use: `ps -e -o args` to see the arguments of the processes). This could prove troublesome since Wazuh processes names make use of extremely common words, leading to potential incorrect process tracking and waste of resources. For example, when looking for processes with the word "db" on it in my machine I get 19 results (1 is the grep itself):
```shell
$ ps -e -O args | grep db | wc -l
20
```
Some examples of processes containing the word "db":
```
PID  COMMAND                        S ?        TIME       ARGS
---  ------------------------------ - -------- ---------- ------------------------------------
###  /usr/lib/systemd/systemd-us     S ?        00:00:00   /usr/lib/systemd/systemd-userdbd
###  /usr/bin/dbus-broker-launch     S ?        00:00:00   /usr/bin/dbus-broker-launch --scope system --audit
###  dbus-broker --log # --contr     S ?        00:00:00   dbus-broker --log # --controller # --machine-id XXXXXXX --max-bytes ######## --max-fds ###### --max-matches ######## --audit
###  /usr/bin/dbus-broker-launch     S ?        00:00:00   /usr/bin/dbus-broker-launch --scope user
###  dbus-broker --log # --contr     S ?        00:00:00   dbus-broker --log # --controller # --machine-id XXXXXXX --max-bytes ######## --max-fds ###### --max-matches ########
###  /usr/bin/gmenudbusmenuproxy     S ?        00:00:02   /usr/bin/gmenudbusmenuproxy
```

- Is it worth the effort and time? In my opinion, no. Here are some points to support my opinion:
  - It seems like this script was made for a very specific task and currently there doesn't seem to be any prospect of being used in other scenarios.
  - Making a real "generic/universal" process resource usage tracking program would take a considerable amount of time since there are many scenarios to be taken into account, and let's not forget that it would have to support several OS (Linux, Windows and MacOS in the very least, since all of these ones can have and agent installed, but so do Solaris, AIX, HP). Even though Python aims to be an OS-independent programming language, it can't avoid it (see the conditional in the `monitor.py` module).
  - Due to the ever-changing nature of software development it is virtually impossible to make a product that is completely future-proof. Pursuing this goal may result in a huge waste of time and resources, since it would have to be modified in the future anyway, either for maintenance or improvements.
  - After saying all this it seems feasible to me that modifying the current script and/or making a new script  from scratch whenever it is needed, could be less time-consuming than the alternative.
  - Other problems that arise from making a full-blown program is that new concerns would carry new concerns to take into account, such as performance, stability, etc.

Overall, it does not seem worth it (at least for now). This is just my honest opinion, suggestions are welcome.